### PR TITLE
Functional aws_iam_policy resource

### LINKED ISSUE #179: aws_iam_policy resource
Create a skeletal aws_iam_policy resource, singular, so we can look up policies by name.

CIS 1.22

### LINKED ISSUE #180: Add attached_entities properties to aws_iam_policy
For CIS 1.22 .

If it makes sense / is possible, split into attached_users / groups / roles.

### PR DESCRIPTION
**Pull request by rx294**
Fixes #179
Fixes #180

Signed-off-by: Rony Xavier <rx294@nyu.edu>

### CODE REVIEW

File: test/unit/resources/aws_iam_policy_test.rb [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      policy['arn:aws:iam::aws:policy/test-policy-2'] =
+      {
+        policy_groups: [],
+        policy_users: [],
+        policy_roles: [],
+      }
+      OpenStruct.new( policy[query[:policy_arn]] )
+    end
+  end
+end
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_iam_policy_test.rb]: @rx294, this is one of the cleanest, most thorough unit test files I have ever seen.  Excellent work!

File: test/integration/default/verify/controls/aws_iam_policy.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  end
+end
+
+control "aws_iam_policy matchers" do
+  describe aws_iam_policy("AWSSupportAccess") do
+    it { should be_attachable }
+  end
+
+  describe aws_iam_policy("AWSSupportAccess") do
+    it { should be_attached }
```
Comment (clintoncwolfe) [inline test/integration/default/verify/controls/aws_iam_policy.rb]: Unless we specifically set it up to be attached in the Terraform code, I don't think we can rely on this test passing, at least not for the specified policy.

File: test/integration/default/verify/controls/aws_iam_policy.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  end
+end
```
Comment (clintoncwolfe) [inline test/integration/default/verify/controls/aws_iam_policy.rb]: Very good integration test file overall.  In the future we should probably add TF code to create a customer-managed policy and attach various things, but for this PR, this is good enough as-is.

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Excellent work overall.  The test coverage is especially good, exemplary even.  There are some clarifications for the docs, and a few implementation issues in the main code.  Many thanks for your contribution!

**Comment by rx294** [comment]
@clintoncwolfe Thank you for the great review...lot of credit goes to @samcornwell whose style i followed from his vpc resource PR. I will make the updates shortly.

File: docs/resources/aws_iam_policy.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
+---
+title: About the aws_iam_policy Resource
+---
+
+# aws_iam_policy
+
+Use the `aws_iam_policy` InSpec audit resource to test properties of a single managed AWS IAM Policy.
+
+To test properties of all managed policies that are available in your AWS account, including your own customer-defined managed policies and all AWS managed policies.
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+A policy is an entity in AWS that, when attached to an identity or resource, defines their permissions. AWS evaluates these policies when a principal, such as a user, makes a request. Permissions in the policies determine whether the request is allowed or denied. 
+
+Each IAM Policy is uniquely identified by its policy_name and arn.
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+<br>
+
+## Syntax
+
+An `aws_iam_policy` resource block identifies a policy by policy name.
+
+    # Find a policy by name
+    describe aws_iam_policy('AWSSupportAccess') do
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+      it { should exist }
+    end
+
+    # Hash syntax for ID
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+The following examples show how to use this InSpec audit resource.
+
+### Test that a policy does exist
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      it { should exist }
+    end
+
+### Test that a policy is attached to entities
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      it { should be_attached }
+    end
+
+### Test that a policy is attachable
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      it { should be_attachable }
+    end
+
+<br>
+
+## Matchers
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+This InSpec audit resource has the following special matchers. For a full list of available matchers (such as `exist`) please visit our [matchers page](https://www.inspec.io/docs/reference/matchers/).
+
+### be_attached
+
+The test will pass if the identified policy is attached to entities.
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      it { should be_attached }
+    end
+
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: Great point!

File: docs/resources/aws_iam_policy.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+### be_attachable
+
+The test will pass if the identified policy is attachable.
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      it { should be_attachable }
+    end
+
+## Properties
+
+### arn
+
+The arn value of the specified policy.
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+### attachment_count
+
+The count of attached entities for the specified policy.
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      its('attachment_count') { should cmp 1 }
+    end
+
+### is_attachable
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: docs/resources/aws_iam_policy.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+The is_attachable bool value of the specified policy.
+
+    describe aws_iam_policy('AWSSupportAccess') do
+      its('is_attachable') { should be true }
+    end
+
+### attached_users
```
[the remark this answers is not in the document]
Comment (rx294) [reply docs/resources/aws_iam_policy.md]: 👍

File: libraries/aws_iam_policy.rb [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    end
+  "
+
+  include AwsResourceMixin
+
+  def to_s
+    "Policy #{@policy_name}"
+  end
+
+  [
```
[the remark this answers is not in the document]
Comment (rx294) [reply libraries/aws_iam_policy.rb]: 👍

File: libraries/aws_iam_policy.rb [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  def validate_params(raw_params)
+    validated_params = check_resource_param_names(
+      raw_params: raw_params,
+      allowed_params: [:policy_name],
+      allowed_scalar_name: :policy_name,
+      allowed_scalar_type: String,
+    )
+
+    if validated_params.empty?
+      raise ArgumentError, 'You must provide parameter to aws_iam_policy: policy name.'
```
[the remark this answers is not in the document]
Comment (rx294) [reply libraries/aws_iam_policy.rb]: 👍

File: libraries/aws_iam_policy.rb [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    end
+
+    validated_params
+  end
+
+  def fetch_from_aws
+    backend = AwsIamPolicy::BackendFactory.create
+
+    criteria = { max_items: 1000 } # maxItems max value is 1000
+    resp = backend.list_policies(criteria)
```
[the remark this answers is not in the document]
Comment (rx294) [reply libraries/aws_iam_policy.rb]: cool thanks for the info

File: libraries/aws_iam_policy.rb [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    @policy = resp.policies.select { |policy| policy.policy_name == @policy_name }.first.to_h
+    @exists = !@policy.empty?
+
+    return unless @exists
+    criteria = { policy_arn: arn }
+    resp = backend.list_entities_for_policy(criteria)
```
[the remark this answers is not in the document]
Comment (rx294) [reply libraries/aws_iam_policy.rb]: 👍

File: test/integration/default/verify/controls/aws_iam_policy.rb (hunk 2 above)
Comment (rx294) [reply test/integration/default/verify/controls/aws_iam_policy.rb]: 👍

File: test/integration/default/verify/controls/aws_iam_policy.rb (hunk 3 above)
Comment (rx294) [reply test/integration/default/verify/controls/aws_iam_policy.rb]: Yes thank you, I am yet to get familiar with TF...for the time being i have changed the controls to ones that will pass.

File: test/unit/resources/aws_iam_policy_test.rb (hunk 1 above)
Comment (aaronlippold) [reply test/unit/resources/aws_iam_policy_test.rb]: @rx294 way to go!

**Comment by rx294** [comment]
@clintoncwolfe I have made the updates. Thanks again!

**Review by clintoncwolfe — APPROVED** [review]

**Comment by rx294** [comment]
Thank you Clinton!

**Pull request merged**